sycl/unittests/scheduler/StreamInitDependencyOnHost.cpp

```diff
@@ -0,0 +1,122 @@
+          std::move(CGH->MOSModuleHandle), std::move(CGH->MStreamStorage),
+          CGH->MCGType, CGH->MCodeLoc));
+      break;
+    }
+    default:
+      throw runtime_error("Unhandled type of command group",
+                          PI_INVALID_OPERATION);
+    }
+
+    return CommandGroup;
+  }
+};
+
+using CmdTypeTy = cl::sycl::detail::Command::CommandType;
+
+static bool ValidateDepCommandsTree(const detail::Command *Cmd,
+                                    std::queue<CmdTypeTy> DepCmdsTypes,
+                                    const detail::SYCLMemObjI *MemObj) {
+  if (DepCmdsTypes.empty())
+    return true;
+  else if (!Cmd)
+    return false;
+
+  CmdTypeTy DepCmdType = DepCmdsTypes.front();
+  DepCmdsTypes.pop();
+
+  for (const detail::DepDesc &Dep : Cmd->MDeps) {
+    if (Dep.MDepCommand && (Dep.MDepCommand->getType() == DepCmdType) &&
+        Dep.MDepRequirement && (Dep.MDepRequirement->MSYCLMemObj == MemObj) &&
+        ValidateDepCommandsTree(Dep.MDepCommand, DepCmdsTypes, MemObj))
+      return true;
+  }
+
+  return false;
+}
+
+TEST_F(SchedulerTest, StreamInitDependencyOnHost) {
+  cl::sycl::queue HQueue(host_selector{});
+  detail::QueueImplPtr HQueueImpl = detail::getSyclObjImpl(HQueue);
+
+  // Emulating processing of command group function
+  MockHandler MockCGH(HQueueImpl, true);
+  MockCGH.setType(detail::CG::KERNEL);
+
+  // Emulating construction of stream object inside command group
+  detail::StreamImplPtr StreamImpl =
+      std::make_shared<detail::stream_impl>(1024, 200, MockCGH);
+  detail::GlobalBufAccessorT FlushBufAcc =
+      StreamImpl->accessGlobalFlushBuf(MockCGH);
+  MockCGH.addStream(StreamImpl);
+
+  detail::SYCLMemObjI *FlushBufMemObjPtr =
+      detail::getSyclObjImpl(FlushBufAcc)->MSYCLMemObj;
+  ASSERT_TRUE(!!FlushBufMemObjPtr)
+      << "Memory object for stream flush buffer not initialized";
+
+  unique_ptr_class<detail::CG> MainCG = MockCGH.finalize();
+
+  // Emulate call of Scheduler::addCG
+  vector_class<detail::StreamImplPtr> Streams =
+      static_cast<detail::CGExecKernel *>(MainCG.get())->getStreams();
+  ASSERT_EQ(Streams.size(), 1u) << "Invalid number of stream objects";
+
+  initStream(Streams[0], HQueueImpl);
+
+  MockScheduler MS;
+  detail::Command *NewCmd = MS.addCG(std::move(MainCG), HQueueImpl);
+  ASSERT_TRUE(!!NewCmd) << "Failed to add command group into scheduler";
+  ASSERT_GT(NewCmd->MDeps.size(), 0u)
+      << "No deps appeared in the new exec kernel command";
+
+  // Searching in dependencies for CG execution command that initializes flush
+  // buffer of a stream that is supposed to be used inside NewCmd's CG.
+  // Tree of dependencies should look like:
+  // [MAIN_CG] -> [EMPTY_NODE {FlushBufMemObj}] -> [FILL_CG {FlushBufMemObj}] ->
+  //     [[ALLOC_TASK {FlushBufMemObj}]
+  std::queue<CmdTypeTy> DepCmdsTypes({CmdTypeTy::EMPTY_TASK,
+                                      CmdTypeTy::RUN_CG, // FILL_CG
```

Comment:
Should one check that this RUN_CG is for host task and the dependency here is for memobject of stream?

Reply:
MemObj parameter contains address of stream's flush buffer object. It is checked in ValidateDepCommandsTree function. I'll add comment.

Reply:
Comment is added.